src/psyclone/psyGen.py

```diff
@@ -2447,24 +2458,40 @@ def _get_private_list(self):
         :raises InternalError: if a Kernel has local variable(s) but they \
                                aren't named.
         '''
-        result = []
-        # get variable names from all loops that are a child of this node
-        for loop in self.loops():
-            # We must allow for implicit loops (e.g. in the NEMO API) that
-            # have no associated variable name
-            if loop.variable_name and \
-               loop.variable_name.lower() not in result:
-                result.append(loop.variable_name.lower())
-        # Get variable names from all kernels that are a child of this node
+        result = set()
+        # get variable names from all calls that are a child of this node
         for call in self.kernels():
             for variable_name in call.local_vars():
                 if variable_name == "":
                     raise InternalError(
                         "call '{0}' has a local variable but its "
                         "name is not set.".format(call.name))
-                if variable_name.lower() not in result:
-                    result.append(variable_name.lower())
-        return result
+                result.add(variable_name.lower())
+
+        # Now determine scalar variables that must be private:
+        var_accesses = VariablesAccessInfo()
+        self.reference_accesses(var_accesses)
+        for var_name in var_accesses.get_all_vars():
+            accesses = var_accesses.get_varinfo(var_name).get_all_accesses()
```

Comment:
Rather than having to call `get_varinfo` would it be worth having some sort of shortcut since this is going to be a very common operation. c.f. Sergi's work on the PSyIR, could we implement things such that `var_accesses[var_name]` does what we want?

Reply:
Good idea, done. I also added a __getitem__ for the list of access information.